### PR TITLE
Add prune command

### LINKED ISSUE #173: Add cxflow prune command
With rapid experimenting, one typically creates tens or even hundreds of log dirs with bogus contents. We should provide an effective way to get rid of these dirs. E.g. add `cxflow prune` command which would be able to

- [x] delete (log) sub-directories with incomplete training artifacts (erroneous training typically creates `train.log` but not `trace.yaml`)

- [x] delete training(s) with less than a specified threshold number of epochs

- [x] delete all sub-directories in training directories (e.g.: visual dirs)

### PR DESCRIPTION
**Pull request by luxusko**
resolve #173 

Not sure what to do in case of rmtree throwing errors (e.g.: deleting symlink), letting the user see the error seems like a good option

### CODE REVIEW

File: cxflow/cli/prune.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from os import path, listdir
+from ruamel import yaml
+from shutil import rmtree
+from cxflow.constants import CXF_DEFAULT_LOG_DIR
+
+
+TRACE_FILE = 'trace.yaml'
+EPOCHS_TRACE_ENTRY = 'epochs_done'
+
+def _prune_all(dir_: str) -> None:
```
Comment (blazekadam) [inline cxflow/cli/prune.py]: We decided to drop this feature as its effect can be achieved with `rm -rf log` command.

File: cxflow/cli/prune.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    rmtree(dir_)
+
+def _prune_subdirs(dir_: str) -> None:
+    """
+    Delete all subdirs in training log dirs.
+
+    :param dir_: dir with training log dirs
+    """
+    for logdir in [path.join(dir_, f) for f in listdir(dir_) if path.isdir(path.join(dir_, f))]:
```
Comment (blazekadam) [inline cxflow/cli/prune.py]: Please filter out training dirs only. You can import `is_training_dir` function from `cxflow.cli.ls`.

File: cxflow/cli/prune.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        for subdir in [path.join(logdir, f) for f in listdir(logdir) if path.isdir(path.join(logdir, f))]:
+            rmtree(subdir)
```
[the remark this answers is not in the document]
Comment (blazekadam) [reply cxflow/cli/prune.py]: Also the user should be informed with `logging.debug` every time you delete something.

File: cxflow/cli/prune.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def _prune(dir_: str, epochs: int) -> int:
+    """
+    Delete all short training logs.
```
Comment (blazekadam) [inline cxflow/cli/prune.py]: Delete all training **dirs** with incomplete training artifacts or with less than specified ``epochs`` done.

File: cxflow/cli/prune.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    :param dir_: dir with training log dirs
+    :param epochs: minimum number of finished epochs to keep the training logs
+    :return: number of log dirs pruned
+    """
+    deleted = 0
+    for logdir in [path.join(dir_, f) for f in listdir(dir_) if path.isdir(path.join(dir_, f))]:
+        trace_path = path.join(logdir, TRACE_FILE)
+        if not path.exists(trace_path):
```
Comment (blazekadam) [inline cxflow/cli/prune.py]: Test with `is_training_dir`.

File: cxflow/cli/prune.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        else:
+            trace_file = open(trace_path, 'r')
+            trace = yaml.load(trace_file, yaml.RoundTripLoader)
+            if trace.get(EPOCHS_TRACE_ENTRY, 0) < epochs:
+                deleted += 1
+                rmtree(logdir)
+
+    return deleted
+
+def prune_train_dirs(dir_: str, all_: bool, epochs: int, subdirs: bool) -> None:
```
Comment (blazekadam) [inline cxflow/cli/prune.py]: Import this function in `__init__.py`

File: cxflow/tests/cli/prune_test.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+"""
```
Comment (blazekadam) [inline cxflow/tests/cli/prune_test.py]: Tests! I like tests! :)

File: cxflow/cli/prune.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
+"""
+Module with `cxflow prune` CLI functionality.
+"""
+from os import path, listdir
+from ruamel import yaml
+from shutil import rmtree
+from cxflow.constants import CXF_DEFAULT_LOG_DIR
```
[the remark this answers is not in the document]
Comment (luxusko) [reply cxflow/cli/prune.py]: I don't understand what's the policy on this. [PR](https://github.com/Cognexa/cxflow/pull/185) I refined few days ago had exactly opposite change request.

**Pull request merged**